### ISSUE TITLE
Can't free soa<> type

### ISSUE THREAD

**Issue by cshenton** [ispc/ispc#2202]
`Intel(r) Implicit SPMD Program Compiler (Intel(r) ISPC), 1.16.1 (build date Jul 15 2021, LLVM 12.0.1)`

The following code fails to compile, the call to `delete` results in an invalid bitcast, attempting to compile the following:

```c
struct thing_t {
    float x;
    float y;
};

static soa<16> thing_t * uniform g_things;

export void alloc_then_free_things() {
    g_things = uniform new soa<16> thing_t[1024];
    delete[] g_things;
}
```

Results in the following error message from the compiler.

```
Invalid bitcast
  %ptr_to_void = bitcast { %v8_soa16_thing_t*, i32 } %g_things_load to i8*, !dbg !22789
Invalid bitcast
  %ptr_to_void14 = bitcast { %v8_soa16_thing_t*, i32 } %g_things_load13 to i8*, !dbg !22794
in function alloc_then_free_things___
LLVM ERROR: Broken function found, compilation aborted!
D:\users\aneshlya\beta_update\v1.16.1\src\main.cpp(392): FATAL ERROR: Unhandled signal sent to process; terminating.
***
*** Please file a bug report at https://github.com/ispc/ispc/issues
*** (Including as much information as you can about how to reproduce this error).
*** You have apparently encountered a bug in the compiler that we'd like to fix!
***
```

I am a bit concerned by this. Should I just be avoiding the `soa` type qualifier entirely? Or am I using it incorrectly?

**Comment by cshenton** [comment]
I wouldn't mind passing in memory from the calling C/C++ code, but I can't cast a `void*` to the correct amount of memory to a pointer to (array of) SoA type either.

**Comment by dbabokin** [comment]
`soa<>` has a number of known bugs and generally should be avoided in favor of using just `varying` types.

**Comment by cshenton** [comment]
Okay, thanks for letting me know, I was wondering why Embree and Unreal Engine had no uses of it, and now I know!

Perhaps the documentation should be updated to mark the feature as experimental (or even moved to a dedicated experimental section)? In particular, the performance guide points the user towards the `soa<>` qualifier, which it probably should not. If there's no plan to fix the bugs with `soa<>` perhaps it could just be moved out of mainline to a branch for when there's enough bandwidth to get it fully working?

Otherwise people might get the impression that other language features which _are_ stable are just as buggy. I know for me personally it was a bit of mental whiplash to hit internal errors on my first serious use of a compiler which is trusted by fairly reputable downstream users.

**Comment by dbabokin** [comment]
I agree, we should either fix it / bring to product quality, or make it experimental / explicitly note that it's not recommended to use it. I'm not sure which path to choose here, but probably moving to experimental is the way.

**Comment by aneshlya** [comment]
Related to https://github.com/ispc/ispc/issues/1591.
